### PR TITLE
Add weisfeiler lehman subgraph hashing

### PR DESCRIPTION
**Pull request by aaronzo**
## Changes
- add a new method `weisfeiler_lehman_hash_subgraphs`, and refactored `weisfeiler_lehman_hash_graph` so the two methods share common steps and are more efficient.
- added docstring with example for `weisfeiler_lehman_hash_subgraphs`
- added test class for `weisfeiler_lehman_hash_subgraphs` with randomly generated graphs for each test
- replaced tests for  `weisfeiler_lehman_hash_graph`  to be aligned with the coverage of tests for `weisfeiler_lehman_hash_subgraphs`

## Description
The current `weisfeiler_lehman_hash_graph` returns a hash for a graph, but sometimes it is useful to keep the subgraph hashes induced by k-hops from each node generated by the iterations along the way. For example when building general WL graph kernels [1] or in the popular 'Graph2Vec' algorithm [2] where these subgraph hashes are viewed as 'words' in the graph to build features from. Non-isomorphic graphs may have different graph hashes but may share lots of subgraph hashes, implying that locally the graphs are similar.

As a result I've added the `weisfeiler_lehman_hash_subgraphs` method which returns a dictionary indexed by node of subgraph hash lists in order of iteration depth.  Similarly to `weisfeiler_lehman_hash_graph`, the user can choose to set: initial node labels for the hashing, an edge attribute used in neighborhood aggregations, iteration depth, or hash digest size with optional keyword arguments.

### References
[1] W-L graph kernels https://www.jmlr.org/papers/volume12/shervashidze11a/shervashidze11a.pdf
[2]  Graph2Vec: https://arxiv.org/abs/1707.05005

Possibly of interest to @cgoliver & @rossbar, authors of the original implementation here.

### CODE REVIEW

**Comment by cgoliver** [comment]
Looks really nice @aaronzo ! Let me know if there is anything I can help with.

File: networkx/algorithms/graph_hashing.py
Comment (dschult) [inline networkx/algorithms/graph_hashing.py:8]:
```suggestion
```

File: networkx/algorithms/graph_hashing.py
Comment (dschult) [inline networkx/algorithms/graph_hashing.py:24]:
```suggestion
        return {u: str(val) for u, val in G.nodes(data=node_attr)}
    elif edge_attr:
        return {u: "" for u in G}
    else:
        return {u: str(deg) for u, deg in G.degree()}
```

File: networkx/algorithms/graph_hashing.py
Comment (dschult) [inline networkx/algorithms/graph_hashing.py:35]:
```suggestion
    for nbr in G.neighbors(node):
        prefix = "" if edge_attr is not None else str(G[node][nbr][edge_attr])
        label_list.append(prefix + node_labels[nbr])
```
Many objects map to False in an "if" clause. So safer to check directly for the object "None"

File: networkx/algorithms/tests/test_graph_hashing.py
Comment (dschult) [inline networkx/algorithms/tests/test_graph_hashing.py:14]: Similar comment for this `_relabel_dict` function. Input `f` is not used.  Maybe simple enough to just inline this code.

File: networkx/algorithms/tests/test_graph_hashing.py
Comment (dschult) [inline networkx/algorithms/tests/test_graph_hashing.py:24]: This class object is never used as a class. Turn all these methods into functions and remove the class test structure.  The functions then have no arguments: e.g.   `def test_empty_graph_hash():`

File: networkx/algorithms/tests/test_graph_hashing.py
Comment (dschult) [inline networkx/algorithms/tests/test_graph_hashing.py:249]: This class is similarly not used as a class and should be changed to a set of test functions.

File: networkx/algorithms/tests/test_graph_hashing.py
Comment (dschult) [inline networkx/algorithms/tests/test_graph_hashing.py:287]: It is standard to use `{}` instead of `dict()`.

**Review by dschult — COMMENTED** [review]
Thanks for this!   It looks good!

File: networkx/algorithms/graph_hashing.py
Comment (aaronzo) [reply networkx/algorithms/graph_hashing.py:8]: good catch - unused import

File: networkx/algorithms/tests/test_graph_hashing.py
Comment (aaronzo) [reply networkx/algorithms/tests/test_graph_hashing.py:24]: thanks for this pointer, I had assumed classes in pytest were for grouping tests and used it as such, will change

File: networkx/algorithms/graph_hashing.py
Comment (aaronzo) [reply networkx/algorithms/graph_hashing.py:35]: I think the correct version would be ` edge_attr is None` (inverted logic) but will change to check for None

File: networkx/algorithms/graph_hashing.py
Comment (aaronzo) [reply networkx/algorithms/graph_hashing.py:24]: an unwanted side effect of 
```
{u: str(val) for u, val in G.nodes(data=node_attr)}
```
is that when a node has no node attribute defined, this will give the string `'None'` in this dictionary, whereas previously this raised a `KeyError`, which I think is helpful behaviour for the user (gives them the responsibility for filling default values for node attributes. Otherwise graphs could pass isomorphism tests unexpectedly if enough nodes have missing attributes

File: networkx/algorithms/tests/test_graph_hashing.py
Comment (aaronzo) [reply networkx/algorithms/tests/test_graph_hashing.py]: changed to inline

File: networkx/algorithms/graph_hashing.py
Comment (dschult) [reply networkx/algorithms/graph_hashing.py:13]: While the examples shown on the page you refer to seem to create the instance on one line and then use the `update` method on the next line, it seems from the docs that using the constructor parameter `data` (as you have done) is a cleaner way to go when the `data` is short enough to fit all on one line -- their examples would mostly not fit nicely on one line if they included the string as a `data` parameter.  Anyway, that's my perspective... your way looks good.

File: networkx/algorithms/graph_hashing.py
Comment (aaronzo) [reply networkx/algorithms/graph_hashing.py]: Yeah, it's only on initialising labels too so not a bottleneck - but there's no reason not to use your elegant suggestion, so I've changed it

**Comment by aaronzo** [comment]
@dschult great! What’s the next steps with this?

**Comment by dschult** [comment]
We need a second "approval" (someone else to look at the PR) and that may be slower than usual due to lots of people traveling during August and there are the "August finish-up the summer projects" distractions as well. I would expect someone to take it up at some point in the next 2 weeks. Maybe today! (due to this exchange) we'll see...
Thanks for the gentle nudge!

File: networkx/algorithms/graph_hashing.py
Comment (rossbar) [inline networkx/algorithms/graph_hashing.py:203]: The hash_label function isn't selectable, is it - e.g. it's not a parameter for the function? The description here implies more flexibility than their is, so it might be worth rewording. Not critically important though.

**Review by rossbar — COMMENTED** [review]
This generally looks really good, thanks @aaronzo! Just a couple minor suggestions. I do think we should add a release note about the fact that the WL hashes will change between NX v2.6 and 2.7, but I wanted to make sure that I was understanding the hashes correctly before doing so. Does that make sense?

File: networkx/algorithms/graph_hashing.py
Comment (aaronzo) [reply networkx/algorithms/graph_hashing.py]: I think you're looking at an old commit for this one, this should be fixed

File: networkx/algorithms/graph_hashing.py
Comment (aaronzo) [reply networkx/algorithms/graph_hashing.py:106]: Yes @rossbar you are correct, and well spotted. I had thought this too while developing but didn't question why my hashes were different too much because the tests are thorough.

I've not investigated but this could be to do with the fact we call the blake2b hash each time rather than calling `.update` method. If we are very concerned we could check the unhashed versions of some subgraphs before and after this change (effectively replacing the `_hash_label` function with string concatenation).

I would agree that this isn't too detrimental if there is a release warning. Is the release notes something I can update from this PR?

File: networkx/algorithms/graph_hashing.py
Comment (aaronzo) [reply networkx/algorithms/graph_hashing.py:203]: Well caught, this is a remnant an earlier draft where the `hash_label` function was a parameter the user could set. I eventually decided this was unnecessarily complex (higher order parameters) but this would be something we could add later if there's a need for it.

In the meanwhile I'll change this doc to reflect that this function isn't selectable

File: networkx/algorithms/graph_hashing.py
Comment (rossbar) [reply networkx/algorithms/graph_hashing.py:106]: > I would agree that this isn't too detrimental if there is a release warning. Is the release notes something I can update from this PR?

Yes I think we'd want to have a prominent note in `doc/release/release_dev.rst` about the change, maybe in`Highlights` since that's the top section.

Before we do that though, I just want to see what others think about this. In principle, if we wanted to be *really* safe wrt backwards compatibility, we could wait to merge this for the NX 3.0 release, where breaking compatibility with 2.X wouldn't be as big a concern. IMO this is overkill - this PR contains a nice new feature and I think the backwards incompatible part is unlikely to affect users (the use-case I can think of where this might be an issue is if users are generating graph hashes, saving them to file, then loading them later for comparison). So my two cents: add a prominent note in `release_dev` and put it in for 2.7. Thoughts?

File: networkx/algorithms/graph_hashing.py
Comment (rossbar) [reply networkx/algorithms/graph_hashing.py]: Yeah, sorry about that - I started a review a while ago when this was still relevant and for whatever reason GH still "remembers" my comment. Sorry for the noise!

File: networkx/algorithms/graph_hashing.py
Comment (aaronzo) [reply networkx/algorithms/graph_hashing.py]: yeah, definitely didn't make sense. Changed it to hopefully be clearer

**Comment by aaronzo** [comment]
Thanks @rossbar picked up the minor docstring changes and replied regarding the release note warning on hash changes for v2.7.
Long story short I agree that it's fine as a warning, though we could investigate more as to why the hashes changed .

File: networkx/algorithms/graph_hashing.py
Comment (aaronzo) [reply networkx/algorithms/graph_hashing.py:103]: Yes, a possible usecase would be change detection on a large database of graphs - on upgrading to 2.7 users might find every graph is flagged as being changed unless they get a warning telling them to recompute all hashes first.

File: networkx/algorithms/graph_hashing.py
Comment (dschult) [inline networkx/algorithms/graph_hashing.py:165]: What do you mean by "subgraph originating from each node"?
Are these induced subgraphs? On what set of nodes?

Also, this should be one-line short description of the function.  Actually, in this case, you should make this first sentence the first paragraph of the doc_string. And make a first line that is short (used to label the links in the doc to this function). Something like:  Return as dict of hashes for each subgraph containing each node.

File: networkx/algorithms/graph_hashing.py
Comment (dschult) [reply networkx/algorithms/graph_hashing.py:103]: I'm a little worried that we don't know why the hashes changed. Did we change the order of operations or something like that which might affect the hash?  What is invariant for a hash of a graph? -- certainly the same graph gives the same hash if we do everything in the same order...

File: networkx/algorithms/graph_hashing.py
Comment (aaronzo) [reply networkx/algorithms/graph_hashing.py:165]: on each iteration, each node label is replaced with the hash of its 1-hop neighbors labels concatenated. So, at iteration k the hash each node label is influenced by all nodes in a radius of 2*k. So, the hash at iteration i for each node n is the WL hash for the subgraph induced by nodes in a radius 2i around n.

It's really hard to put it succinctly but accurately, I feel like english language can handle dicts and lists but plurals start to get confusing when talking about a dict of lists at each iteration! Suggestions welcome!

Comment (aaronzo) [reply networkx/algorithms/graph_hashing.py:165]: how about:
```
Return a dictionary of subgraph hashes by node. The hashed subgraphs
are induced by increasing k-hop neighborhoods around each node, where
k is even.
```
still too long?

File: networkx/algorithms/graph_hashing.py
Comment (aaronzo) [reply networkx/algorithms/graph_hashing.py:103]: Ok, I did a bit of preliminary investigation. I've not found the reason for the hash difference between this and main, _but_ I have found a bug on main which persists in this branch. It comes from the `neighborhood_aggregate` function. This was left unchanged in this PR apart from 
```python
prefix = "" if not edge_attr else G[node][nei][edge_attr]` # old
prefix = "" if edge_attr is None else str(G[node][nei][edge_attr]) # new
```
(which is more robust but not should not cause hash changes to well-behaved inputs) and the fact that it was copied outside the WL-hash function so that both hashing methods could use it.

The issue is that this function (written here simplified without edge attributes at all):
```python
def _neighborhood_aggregate(G, node, node_labels):
    label_list = [node_labels[node]]
    for nbr in G.neighbors(node):
        label_list.append(node_labels[nbr])
    return "".join(sorted(label_list))
```
makes no distinction between the original node and its neighbors, since sorting combines both.

This I'm pretty sure is bad behaviour. The simplest counterexample I could come up with was the following:

```python
G = nx.DiGraph()
G.add_nodes_from([
   (1, {'node_attr': 'A'}),
   (2, {'node_attr': 'B'}),
   (3, {'node_attr': 'C'})
])
G.add_edges_from([
   (1, 2),
   (2, 3),
   (3, 1)
])
```
i.e. G is a directed cyclic triangle. Then under the current implementation 
```python
weisfeiler_lehman_graph_hash(G, node_attr='node_attr')
weisfeiler_lehman_graph_hash(G2.reverse(), node_attr='node_attr')
```
the above hashes are equal at any depth, i.e `G` and `G.reverse()` pass an isomorphism test, even though G has an edge from node type A -> B and its reverse does not.

It's probably possible to come up with 2 non-isomorphic undirected graphs which pass due to this technicality.

I believe using instead
```python
def _neighborhood_aggregate(G, node, node_labels):
    label_list = []
    for nbr in G.neighbors(node):
        label_list.append(node_labels[nbr])
    return str(node_labels[node]).join(sorted(label_list))
```
would catch these cases, but certainly this would cause graph hashes to change across the board, requiring the warning. This should definitely be included though, since we have graphs the algorithm is flagging as isomorphic when it should be able to discern they are not (in general the algorithm does not imply isomorphism, just gives a high likelihood in practice)

FYI I caught this by changing the `_hash_label` function to just return the label as an unhashed string, this way the final graph `hash` for the above graph was (after 1 iteration):
```
(('AB', 1), ('AC', 1), ('BC', 1))
```
which is much easier to debug than a random hash.

I'm still going to try and compare with main to see if any operations are swapped regardless of this bug.

File: networkx/algorithms/graph_hashing.py
Comment (aaronzo) [reply networkx/algorithms/graph_hashing.py:165]: You get a hash per iteration, so for each node you get a list in increasing iteration order of hashes.
Maybe the following?
```
Return a dictionary where each key is a node in the graph and each value is a list containing hashes of increasing size subgraphs containing that node. 

The hashed subgraphs are induced on neighbors within k edges of the node, for each even value of k.
```

Comment (dschult) [reply networkx/algorithms/graph_hashing.py:165]: I like the second option because it has a nice short first line the docs can use for links to this function.
The word "even" gets kind of lost in the last sentence. And how do we know when to stop increasing k?
What do you think of:
```
Return a dictionary of subgraph hashes by node. 

The dictionary is keyed by node to a list of hashes in increasingly
sized induced subgraphs containing the nodes within 2*k edges
of the key node for increasing integer k until all nodes are included.
```

Comment (aaronzo) [reply networkx/algorithms/graph_hashing.py:165]: I think you've cracked it!

File: networkx/algorithms/graph_hashing.py
Comment (aaronzo) [reply networkx/algorithms/graph_hashing.py:103]: I've resolved why the hashes were different - in fact it's a bug I fixed early in development and forgot to mention in the description. I'll make a separate comment thread summarizing.

**Comment by aaronzo** [comment]
This PR in addition to adding subgraph hashes fixes 2 bugs in the original `weisfeiler_lehman_graph_hash` function:

1.  At each iteration depth, the function previously updated node labels by sorting and then concatenating labels in a 1-hop neighborhood. The labels were then hashes and a counter of hashes updated, **but the labels themselves were not updated with the hashed version(*, simply the concatenated string. This leads to exponential growth in label lengths stored and complexity in concatenating them. With `depth=20` and a simple graph I could get the algorithm to run indefinitely before producing a `MemoryError`. At depth 18 the the original implementation took ~30000ms whilst the implementation in this pull request took 3ms. The relevant changes are in [this section of the PR ](https://github.com/networkx/networkx/pull/4946/files#diff-4cb7a313b6d611748e83478406a94c69a5ba10a6cb97e4f2607a61dc4056c72aL120-R143). I ran my tests on the graph below, found in the docstrings for `weisfeiler_lehman_graph_hash`:
```
G = nx.Graph()
G.add_edges_from(
    [
        (1, 2, {"edge_attr": "A"}),
        (2, 3, {"edge_attr": "A"}),
        (3, 1, {"edge_attr": "A"}),
        (1, 4, {"edge_attr": "B"}),
    ]
)
```

2. Originally, when neighborhoods were aggregated through sorting and concatenating, there was [no distinction made between the subject node label and its neighbors labels](https://github.com/networkx/networkx/pull/4946/files#diff-4cb7a313b6d611748e83478406a94c69a5ba10a6cb97e4f2607a61dc4056c72aL110). This has the effect that directed graphs with the edges reversed wrongly pass the WL isomorphism test. A more detailed explanation [can be found here](https://github.com/networkx/networkx/pull/4946#discussion_r700689923). This bug is not exclusive to directed graphs, it applies whenever a graph has 2 distinct neighborhoods (containing the subject node) which have distinct nodesets - for example incomplete graphs. Using a an `edge_attr` parameter would have avoided this bug in the past since only neighbors were prefixed with their connecting edge attributes at the aggregation stage. In this PR, the subject node is [always first in the concatenated string](https://github.com/networkx/networkx/pull/4946/files#diff-4cb7a313b6d611748e83478406a94c69a5ba10a6cb97e4f2607a61dc4056c72aR35), fixing this issue

Both of these fixes cause the values of hashes to change between version 2.6 and 2.7, so there's a warning to be added under the release notes ([see earlier discussion](https://github.com/networkx/networkx/pull/4946#pullrequestreview-703011180))

File: networkx/algorithms/graph_hashing.py
Comment (aaronzo) [reply networkx/algorithms/graph_hashing.py]: updated this, hopefully it's clear enough now

**Comment by aaronzo** [comment]
The failing checks are pending the merge of #5068 only AFAICT. Other than that I think I've resolved everything @rossbar @dschult, apart from updating `release_dev.rst`

File: doc/release/release_dev.rst
Comment (aaronzo) [inline doc/release/release_dev.rst:28]: @rossbar @dschult added a draft of a warning to `release_dev.rst` on hash changes. Feel free to rewrite this. I think I've resolved all other issues in the PR.

**Comment by dschult** [comment]
Very Nice!
Thanks!

**Comment by rossbar** [comment]
Thanks @aaronzo . I've made a few updates to the release note (and fixed a sphinx warning) and was attempting to push the changes up to your branch but it looks like that's been disabled. If you're comfortable doing so, you can check the "allow edits from maintainers" button on the right and I will push up changes. If you're not comfortable with that (or it's not working for whatever reason), just LMK and I'll open a PR to your fork instead.

**Comment by aaronzo** [comment]
@rossbar apologies I've enabled that, go ahead

**Review by rossbar — APPROVED** [review]
Thanks @aaronzo , this LGTM and thanks for chasing down what was going on with the original WL hashing function. I feel much better now that the difference in the hashes has been explained.

The release note looks good, I tweaked it a bit mostly for RST formatting but I did change the wording a bit so please double check that it makes sense!

**Comment by aaronzo** [comment]
Had a look, makes sense @rossbar thanks for the touch-up. Ready to merge.

**Comment by rossbar** [comment]
@dschult just a quick double-check that your approval still stands since quite a bit has changed since the original approval. If so, we can go ahead an put this in but if you (or anyone else) would like another look that's of course totally fine!

**Pull request merged**
